### PR TITLE
R2: use fraction of deviance explained for nonnegative views

### PR DESCRIPTION
**Pull request by ilia-kats**
R2 as a measure of variance explained is not well defined for non-negative views, where the null model predicts the minimum of the data. It also doesn't even work at all for negative binomial views, since the R2 calculation contains a division by the data mimimum, which will be 0 in most cases.

Instead, use the fraction of deviance explained for nonnegative views. It reduces to R2 for Gaussian likelihoods and has a sufficiently similar interpretation.

### CODE REVIEW

File: src/mofaflex/_core/likelihoods/negativebinomial.py [hunk 1]
```diff
@@ -80,6 +82,33 @@ def _r2_impl(
...
+        variance = np.nanvar(y_true, axis=0, mean=truemean)
+
+        # use estimated dispersion for saturated model
+        # naive variance is most likely overestimated since the minimum of the data is used instead of the mean
+        loglik_saturated = nbinom.logpmf(
+            y_true,
+            p=1 / (1 + self._dispersion.mean[feature_idx] * y_true),
+            n=1 / (self._dispersion.mean[feature_idx] + settings.eps),
+        ).sum()
+        loglik_null = nbinom.logpmf(y_true, p=truemean / variance, n=truemean**2 / (variance - truemean)).sum()
```
Comment (arberqoku) [inline src/mofaflex/_core/likelihoods/negativebinomial.py]: For nonnegative views `_shift` is `nanmin(adata.X / sample_means, axis=0)`, which is `0` for essentially every feature with at least one zero count. That gives `p=0, n=0`, so this fails silently as `R² = 0`?
Also, `variance < truemean` gives `n < 0` (`nan`), and `variance == truemean` divides by `0`. And, do we need `nansum` instead of `sum()` or are there no missing values here?

File: src/mofaflex/_core/likelihoods/bernoulli.py [hunk 2]
```diff
@@ -61,6 +62,22 @@ def _r2_impl(
...
+        y_true: Matrix[np.number],
+        y_pred: Matrix[np.floating],
+        group_name: str,
+        sample_idx: Vector[int] | slice = slice(None),
+        feature_idx: Vector[int] | slice = slice(None),
+    ) -> LogLikelihoods:
+        truemean = expit(self._shift[group_name][feature_idx])
+
+        loglik_null = bernoulli.logpmf(y_true, truemean).sum()
+        loglik_model = bernoulli.logpmf(y_true, y_pred).sum()
```
Comment (arberqoku) [inline src/mofaflex/_core/likelihoods/bernoulli.py]: Should `y_pred` be clipped? `transform_prediction` returns `expit(prediction + shift)`, which saturates to exactly 1.0 around an argument of 40, `bernoulli.logpmf(0, 1.0)` is `-inf`, resulting silently to 0.

Comment (ilia-kats) [reply src/mofaflex/_core/likelihoods/bernoulli.py]: Excellent point. I've re-implemented this to calculate logpmf directly from the logits using the logsumexp trick, which should prevent overflow.

File: src/mofaflex/_core/likelihoods/negativebinomial.py (hunk 1 above)
Comment (ilia-kats) [reply src/mofaflex/_core/likelihoods/negativebinomial.py]: Good point. I'm now clipping this to `settings.eps`, but I'm open to better suggestions. I've also switched to nansum everywhere, it slipped my mind that `y_true` might have missing observations.

File: src/mofaflex/_core/likelihoods/base.py [hunk 3]
```diff
@@ -226,15 +263,26 @@ def r2(
...
+                feature_idx,
+            )
+        else:
+            r2 = self._deviance_explained_impl(y_true, y_pred, group_name, sample_idx, feature_idx)
+            if isinstance(r2, LogLikelihoods):
+                r2 = R2(r2.saturated - r2.model, r2.saturated - r2.null)
+
+        with np.errstate(invalid="ignore", divide="ignore"):
+            # np.max, in contrast to max, preserves NaNs, which the caller warns about
+            return np.max((0.0, 1.0 - np.divide(r2.ss_res, r2.ss_tot)))
```
Comment (ilia-kats) [inline src/mofaflex/_core/likelihoods/base.py]: why `np.divide` instead of just `/`?

File: src/mofaflex/_core/model.py [hunk 4]
```diff
@@ -353,7 +355,11 @@ def r2_wrapper(view, group_name, view_name):
                             )
                             for component_name, component in component_iter
                         }
-                if r2_full < settings.eps:
+                if np.isnan(r2_full):
+                    _logger.warning(
+                        f"R2 for view {view_name} could not be calculated, the null model has zero deviance."
```
Comment (ilia-kats) [inline src/mofaflex/_core/model.py]:
```suggestion
                        f"R2 for view {view_name} could not be calculated. Verify your data and model configuration or try to create a minimal reproducible example and report an issue."
```
There may be other sources of `NaN`, not just the null model.
